core/blockchain.go

```diff
@@ -2018,6 +2029,105 @@ func (bc *BlockChain) update() {
+// transaction index.
+//
+// User can use flag `txlookuplimit` to specify a "recentness" block, below
+// which ancient tx indices get deleted. If `txlookuplimit` is 0, it means
+// all tx indices will be reserved.
+//
+// The user can adjust the txlookuplimit value for each launch, Geth will
+// automatically construct the missing indices and delete the extra indices.
+func (bc *BlockChain) updateTxIndices() {
+	var (
+		done   chan struct{} // Non-nil if background unindexing or reindexing routine is active.
+		headCh = make(chan ChainHeadEvent)
+	)
+	sub := bc.SubscribeChainHeadEvent(headCh)
+	defer func() {
+		if sub != nil {
+			sub.Unsubscribe()
+		}
+	}()
+	// initialiseIndices inits txlookup indices into the database.
+	// If there already exists some indices, this function will find
+	// the oldest block which has been indexed and start indexing from
+	// this point.
+	initialiseIndices := func(head uint64) {
+		defer func() { done <- struct{}{} }()
+
+		from, to := uint64(0), head
+		if bc.txLookupLimit != 0 && head > bc.txLookupLimit {
+			from = head - bc.txLookupLimit
+		}
+		// Find oldest indexed block via binary search when we don't
+		// have this flag in database.
+		start := time.Now()
+		oldest := rawdb.FindOldestIndexedBlock(bc.db, from, to)
+		log.Debug("Find oldest indexed block", "oldest", oldest, "elapsed", common.PrettyDuration(time.Since(start)))
+
+		// Re-construct missing tx indices.
+		if oldest == nil {
+			rawdb.IndexTxLookup(bc.db, from, to) // No block has been indexed.
+		} else {
+			rawdb.IndexTxLookup(bc.db, from, *oldest)
+
+			// Drop all useless tx indices below the HEAD-limit.
+			if from > 0 {
+				oldest := rawdb.FindOldestIndexedBlock(bc.db, 0, from)
+				if oldest != nil {
+					rawdb.RemoveTxsLookup(bc.db, *oldest, from)
+				}
+			}
+		}
+		log.Debug("Initialised transaction indices", "elapsed", common.PrettyDuration(time.Since(start)))
+	}
+	// indexBlocks reindex or unindex transaction indices depends
+	// on user's requirement.
+	indexBlocks := func(oldest uint64, head uint64) {
+		defer func() { done <- struct{}{} }()
+
+		// All indices should be reserved.
+		if bc.txLookupLimit == 0 || head <= bc.txLookupLimit {
+			if oldest == 0 {
+				// Short circuit if nothing to delete.
+				return
+			} else {
+				// Reindex all indices if necessary
+				rawdb.IndexTxLookup(bc.db, 0, oldest)
+				return
+			}
+		}
+		if head-bc.txLookupLimit < oldest {
+			// Reindex a part of missing indices and rewind oldest indexed
+			// point to HEAD-limit
+			rawdb.IndexTxLookup(bc.db, head-bc.txLookupLimit, oldest)
+		} else {
+			// Unindex a part of stale indices and forward oldest indexed
+			// point to HEAD-limit
+			rawdb.RemoveTxsLookup(bc.db, oldest, head-bc.txLookupLimit)
+		}
+	}
+
+	for {
+		select {
+		case head := <-headCh:
+			if done == nil {
+				done = make(chan struct{})
+				if number := rawdb.ReadOldestIndexedBlock(bc.db); number == nil {
+					go initialiseIndices(head.Block.NumberU64())
```

Comment:
Lets pass the `done` channel as parameter and only declare done before this for loop. The code will be a lot more readable since we don't have the done channel leaking as a closure into the methods.